Milestone 1.6 and 1.7 (Migrate away from protractor): Migrated few test suite from protractor to webdriverIO.

ShivamJhaa opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

Migrate the following test suite from protractor to webdriverIO.
**Suites Covered:**
* ClassroomPage
* ClassroomPageFileUploadFeatures
* TopicAndSkillDashboard
* SkillEditor
* Wipeout
* Navigation

**Utils Covered:**
* ClassroomPage (Complete)
* LibraryPage (Partial)
* TopicEditorPage (Complete)
* TopicsAndSkillsDashboardPage (Complete)
* AdminPage (Partial)
* ExplorationEditorPage (Partial)
* SkillEditorPage (Complete)
* DeleteAccountPage (Complete)
* GetStartedPage (Complete)



## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The linter/Karma presubmit checks have passed locally on your machine.
- [ ] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [ ] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct
* ClassroomPage
Local: https://drive.google.com/file/d/1NS3LKVRMxGTXlzGJ5BuEESCquGbee6hQ/view?usp=sharing
Cl:
https://github.com/oppia/oppia/runs/7532072176?check_suite_focus=true
* ClassroomPageFileUploadFeatures
Local: https://drive.google.com/file/d/1lLybxEEIjUiWDQ5c3Dct63XZ68Xgt1wK/view?usp=sharing
Cl:
https://github.com/oppia/oppia/runs/7533078108?check_suite_focus=true
* TopicAndSkillDashboard
Local: https://drive.google.com/file/d/1G5CgnMhSAGS6LvQDqBW0VXtRYkWHBkWU/view?usp=sharing
Cl:
https://github.com/oppia/oppia/runs/7532996828?check_suite_focus=true
* SkillEditor
Local: https://drive.google.com/file/d/1Omaw7i5KVSg7q5O9THFFLQl4yKCPGiVe/view?usp=sharing
Cl:
https://github.com/oppia/oppia/runs/7532072430?check_suite_focus=true
* Wipeout
Local: https://drive.google.com/file/d/1JtBGFgr0Cn3hnN2WsoDZiIxs8bIROFAW/view?usp=sharing
Cl:
https://github.com/oppia/oppia/runs/7532072392?check_suite_focus=true
* Navigation
Local: https://drive.google.com/file/d/1l_zrgS65sTh7WfKYV6HvIvD16vk_NkoP/view?usp=sharing
Cl:
https://github.com/oppia/oppia/runs/7532072176?check_suite_focus=true
<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof
for the changes then please leave a comment "No proof of changes needed because {{Reason}}"
and remove all the sections below.
-->


## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

ShivamJhaa committed:
Replace protractor class names with e2e-class name

ShivamJhaa committed:
Reverted unwanted changes

ShivamJhaa committed:
Lint checks

ShivamJhaa committed:
Lint checks

ShivamJhaa committed:
Fixed few backend coverage

ShivamJhaa committed:
Lint checks

ShivamJhaa committed:
Fix merge conflicts

ShivamJhaa committed:
Migrated subscription creatorDashboard preferences

ShivamJhaa committed:
Testing element selector placed at top

ShivamJhaa committed:
Testing element selector placed at top

ShivamJhaa committed:
Fixed element selctor at top not working for AdminPage

ShivamJhaa committed:
Resolved review comments

ShivamJhaa committed:
Resolved review comments

ShivamJhaa committed:
Few backend test fix

ShivamJhaa committed:
Migrated learnerDashboard file

ShivamJhaa committed:
Fix

ShivamJhaa committed:
Resolved review comments

ShivamJhaa committed:
Fixed few backend tests

ShivamJhaa committed:
lint checks

ShivamJhaa committed:
Reverted few change

ShivamJhaa committed:
Fixed all backend tests

ShivamJhaa committed:
Fixed backend coverage

ShivamJhaa committed:
Lint checks

ShivamJhaa committed:
Fix failing e2e

ShivamJhaa committed:
Deleted protractor dependencies

ShivamJhaa committed:
Merge branch 'Milestone-1.2' of https://github.com/ShivamJhaa/oppia into Milestone-1.3

ShivamJhaa committed:
Migrated creatorDashboard

ShivamJhaa committed:
Lint checks

ShivamJhaa committed:
Lint checks

ShivamJhaa committed:
Few fixes

ShivamJhaa committed:
Few fixed

ShivamJhaa committed:
Fixed merge conflicts

ShivamJhaa committed:
lints

ShivamJhaa committed:
lints

ShivamJhaa committed:
comments resolved

ShivamJhaa committed:
fix

ShivamJhaa committed:
Fixed failing tests

ShivamJhaa committed:
Merge branch 'Milestone-1.2' of https://github.com/ShivamJhaa/oppia into Milestone-1.3

ShivamJhaa committed:
Fix

ShivamJhaa committed:
testing add to play later

ShivamJhaa committed:
testing add to play later

ShivamJhaa committed:
Fixed add to play later issue

ShivamJhaa committed:
Fixed merge conflicts

ShivamJhaa committed:
Fixed merge conflicts

ShivamJhaa committed:
few more files migrated

ShivamJhaa committed:
Migrated learnerDashboard tests

ShivamJhaa committed:
Deleted protractor files and dependencies

ShivamJhaa committed:
testing alert issue

ShivamJhaa committed:
Fix merge conflicts

ShivamJhaa committed:
Fix merge conflicts

ShivamJhaa committed:
Migrated profilemenuflow

ShivamJhaa committed:
Fixed failing tests

ShivamJhaa committed:
Fixed failing tests

ShivamJhaa committed:
Merge branch 'develop' of https://github.com/oppia/oppia into Milestone-1.2

ShivamJhaa committed:
few more files migrated

ShivamJhaa committed:
Comments resolved

ShivamJhaa committed:
changed lint rules

ShivamJhaa committed:
Merge conflicts

ShivamJhaa committed:
Checking lint

ShivamJhaa committed:
Checking lint

ShivamJhaa committed:
e2e lints

ShivamJhaa committed:
e2e lints

ShivamJhaa committed:
e2e lints

ShivamJhaa committed:
Sorted imports

ShivamJhaa committed:
lints

ShivamJhaa committed:
lints

ShivamJhaa committed:
testing alert issue

ShivamJhaa committed:
Resolving comments

ShivamJhaa committed:
Resolving comments

ShivamJhaa committed:
testing

ShivamJhaa committed:
Added functions for all selectors

ShivamJhaa committed:
testing

ShivamJhaa committed:
test

ShivamJhaa committed:
Fixed backend tests

ShivamJhaa committed:
Added functions for all selectors

ShivamJhaa committed:
testing

ShivamJhaa committed:
testing

ShivamJhaa committed:
Added action commands in extesnions

ShivamJhaa committed:
Added comments for user.js

ShivamJhaa committed:
Added lint checks

ShivamJhaa committed:
lints

ShivamJhaa committed:
minor fix

ShivamJhaa committed:
nits

ShivamJhaa committed:
Merge branch 'develop' of https://github.com/oppia/oppia into Milestone-1.3

ShivamJhaa committed:
Merge branch 'Milestone-1.2' of https://github.com/ShivamJhaa/oppia into Milestone-1.3

ShivamJhaa committed:
nits

ShivamJhaa committed:
updated deafult time

ShivamJhaa committed:
updated deafult time

ShivamJhaa committed:
Merge branch 'develop' of https://github.com/oppia/oppia into Milestone-1.2

ShivamJhaa committed:
testing

ShivamJhaa committed:
Fixed lint for wdio.conf.js

ShivamJhaa committed:
Merge branch 'develop' of https://github.com/oppia/oppia into Milestone-1.2

ShivamJhaa committed:
lint

ShivamJhaa committed:
merge conflicts

ShivamJhaa committed:
Merge branch 'Milestone-1.3' of https://github.com/ShivamJhaa/oppia into Milestone-1.4

ShivamJhaa committed:
Fixed merge conflicts

ShivamJhaa committed:
comments

ShivamJhaa committed:
nits

ShivamJhaa committed:
Fixed comments

ShivamJhaa committed:
fixed merge conflicts

ShivamJhaa committed:
fixes

ShivamJhaa committed:
Merge branch 'develop' of https://github.com/oppia/oppia into Milestone-1.3

ShivamJhaa committed:
Placed selectors at top

ShivamJhaa committed:
fix

ShivamJhaa committed:
merge comflicts

ShivamJhaa committed:
merge comflicts

ShivamJhaa committed:
fix failing tests

ShivamJhaa committed:
fix failing tests

ShivamJhaa committed:
fix failing tests

ShivamJhaa committed:
comments

ShivamJhaa committed:
fix failing tests

ShivamJhaa committed:
comments resolevd

ShivamJhaa committed:
Merge branch 'develop' of https://github.com/oppia/oppia into Milestone-1.3

ShivamJhaa committed:
Merge branch 'Milestone-1.3' of https://github.com/ShivamJhaa/oppia into Milestone-1.4

ShivamJhaa committed:
Migrated topicAndStoryViewer suite

ShivamJhaa committed:
Deleted protractor version

ShivamJhaa committed:
nit

ShivamJhaa committed:
Resolved comments

ShivamJhaa committed:
Migrated few other dependencies

ShivamJhaa committed:
trying fix embedding error

ShivamJhaa committed:
Migrated ExplorationEditor Dependencies

ShivamJhaa committed:
trying fix embedding error

ShivamJhaa committed:
Reverted few changes

ShivamJhaa committed:
Reverted few changes

ShivamJhaa committed:
Reverted few changes

ShivamJhaa committed:
Fixed all test

ShivamJhaa committed:
Added artifacts upoad code

ShivamJhaa committed:
Migrated topicAndStoryEditorFileUploadFeatures

ShivamJhaa committed:
nits

ShivamJhaa committed:
Migrated blog dashboard test suite

ShivamJhaa committed:
Merge branch 'develop' of https://github.com/oppia/oppia into Milestone-1.3

ShivamJhaa committed:
Merge branch 'Milestone-1.3' of https://github.com/ShivamJhaa/oppia into Milestone-1.4

ShivamJhaa committed:
Merge conflicts

ShivamJhaa committed:
Few fixes

ShivamJhaa committed:
few fixes

ShivamJhaa committed:
nits

ShivamJhaa committed:
nits

ShivamJhaa committed:
updated codeowners

ShivamJhaa committed:
Updated codeowners

ShivamJhaa committed:
Fixed blog dashboard tests

ShivamJhaa committed:
Merge conflicts

ShivamJhaa committed:
Merge conflicts

ShivamJhaa committed:
Merge conflicts

ShivamJhaa committed:
Merge branch 'develop' of https://github.com/oppia/oppia into Milestone-1.5

ShivamJhaa committed:
Merge branch 'develop' of https://github.com/oppia/oppia into Milestone-1.4

ShivamJhaa committed:
Few fixes for contributorDashboard

ShivamJhaa committed:
nits

ShivamJhaa committed:
nits

ShivamJhaa committed:
nits

ShivamJhaa committed:
nits

ShivamJhaa committed:
nits

ShivamJhaa committed:
fix

ShivamJhaa committed:
nits

ShivamJhaa committed:
Merge branch 'develop' of https://github.com/oppia/oppia into Milestone-1.4

ShivamJhaa committed:
nits

ShivamJhaa committed:
fix

ShivamJhaa committed:
comments resolved

ShivamJhaa committed:
check chapter navigation

ShivamJhaa committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

ShivamJhaa committed:
testing learnerdashboard flake

ShivamJhaa committed:
testing learnerdashboard flake

ShivamJhaa committed:
fixes

ShivamJhaa committed:
fixes

ShivamJhaa committed:
fixes

ShivamJhaa committed:
fixes

ShivamJhaa committed:
fixes

ShivamJhaa committed:
Update general.js

ShivamJhaa committed:
Migrated contributor dashboard

ShivamJhaa committed:
Migrated 2 suites

ShivamJhaa committed:
testing

ShivamJhaa committed:
tecting

ShivamJhaa committed:
Mirated topicAndSkillDashboard suite

ShivamJhaa committed:
Migrated all suites

ShivamJhaa committed:
Testing

ShivamJhaa committed:
Comments resolevd

ShivamJhaa committed:
Merged conflicts

ShivamJhaa committed:
Merge branch 'develop' of https://github.com/oppia/oppia into Milestone-1.5

ShivamJhaa committed:
Nits

ShivamJhaa committed:
nits

ShivamJhaa committed:
nits

ShivamJhaa committed:
Nits

ShivamJhaa committed:
merge conflicts

ShivamJhaa committed:
fixes

ShivamJhaa committed:
Deleted protractor depedencies

ShivamJhaa committed:
nits

ShivamJhaa committed:
nits

ShivamJhaa committed:
fix

ShivamJhaa committed:
topic fix

ShivamJhaa committed:
logging level

ShivamJhaa committed:
Testing video reporter

ShivamJhaa committed:
Testing video reporter

ShivamJhaa committed:
nits

ShivamJhaa committed:
check

ShivamJhaa committed:
Resolved merge conflicts

ShivamJhaa committed:
undo some changes

ShivamJhaa committed:
nits

ShivamJhaa committed:
Merge branch 'leanerDashboard' of https://github.com/ShivamJhaa/oppia into Milestone-1.4

ShivamJhaa committed:
Merge branch 'develop' of https://github.com/oppia/oppia into Milestone-1.4

ShivamJhaa committed:
testing

ShivamJhaa committed:
Merge conflicts

ShivamJhaa committed:
Merge branch 'develop' of https://github.com/oppia/oppia into Milestone-1.4

ShivamJhaa committed:
testing

ShivamJhaa committed:
fix

ShivamJhaa committed:
Merge branch 'develop' of https://github.com/oppia/oppia into Milestone-1.5

ShivamJhaa committed:
Resolved comments

ShivamJhaa committed:
fixed failing topicAndStoryEditor test

ShivamJhaa committed:
nits

ShivamJhaa committed:
added classname and comments

ShivamJhaa committed:
nits

ShivamJhaa committed:
testing

ShivamJhaa committed:
nits

ShivamJhaa committed:
fixed merge conflicts

ShivamJhaa committed:
testing

ShivamJhaa committed:
Merge conflicts

ShivamJhaa committed:
merge conflicts

ShivamJhaa committed:
merge conflicts

ShivamJhaa committed:
fixes

ShivamJhaa committed:
fixes

ShivamJhaa committed:
Additional check for navigation test suite

ShivamJhaa committed:
added debug info for waitFor

ShivamJhaa committed:
Merge branch 'Milestone-1.4' of https://github.com/ShivamJhaa/oppia into Milestone-1.5

ShivamJhaa committed:
nits

ShivamJhaa committed:
Merge branch 'Milestone-1.5' of https://github.com/ShivamJhaa/oppia into Milestone-1.7

ShivamJhaa committed:
nits

ShivamJhaa committed:
nits

vojtechjelinek reviewed (approved):
LGTM!

ShivamJhaa committed:
Merge conflicts

ShivamJhaa committed:
Merge confilcts

ShivamJhaa committed:
nits

ShivamJhaa committed:
nits

ShivamJhaa committed:
nits

ShivamJhaa committed:
Comments resolved

kevintab95 reviewed (approved):
LGTM for 2 codeowner files. Thanks!

U8NWXD reviewed (commented):
I'm leaving the e2e files for @SAEb-ai to review, but otherwise LGTM

ShivamJhaa committed:
Fixed comments

ShivamJhaa commented:
@SAEb-ai PTAL, thanks!

ShivamJhaa committed:
fixed merge conflicts

ShivamJhaa committed:
merge conflicts

ShivamJhaa committed:
merge conflicts

ShivamJhaa committed:
nits

SAEb-ai commented on core/tests/webdriverio_desktop/navigation.js:
```diff
@@ -122,35 +121,37 @@ describe('Static Pages Tour', function() {
           return !(/login/.test(url));
-        }, 10000);
+        },
+        {
+          timeout: 10000,
+        });
```
Can you please update the formatting for this as it's occupying too much space?

ShivamJhaa replied:
Done.

SAEb-ai commented on core/tests/webdriverio_desktop/navigation.js:
```diff
@@ -92,28 +92,27 @@ describe('Meta Tags', function() {
 });
 
 describe('Donation flow', function() {
-  var payPalButton = element(by.css('.e2e-test-paypal-donate-button'));
-  var creditCardButton = element(
-    by.css('.e2e-test-credit-card-donate-button'));
```
Why these were removed from here?

ShivamJhaa replied:
As previously discussed we can't place selectors at top in suite files. I am doing same in all suite files that I have migrated.

SAEb-ai replied:
I thought it was meant for selectors outside the describe block.

ShivamJhaa replied:
No it was actually meant for outside the 'it' block. We cannot use $ selector outside the 'it' block as it undefined there.

ShivamJhaa replied:
No it was actually meant for outside the 'it' block. We cannot use $ selector outside the 'it' block as it undefined there.

SAEb-ai commented on core/tests/webdriverio_desktop/topicsAndSkillsDashboard.js:
```diff
@@ -97,6 +97,7 @@ describe('Topics and skills dashboard functionality', function() {
     await topicsAndSkillsDashboardPage.get();
     await topicsAndSkillsDashboardPage.filterTopicsByKeyword(TOPIC_NAME);
     await topicsAndSkillsDashboardPage.expectNumberOfTopicsToBe(1);
+    await topicsAndSkillsDashboardPage.resetTopicFilters();
```
Why this change?

ShivamJhaa replied:
Once we apply the filter, the topic name covers the center space of button that we generally click to apply filter and as we know that the click in webdriverio is always clicked at center of elements, the buttons becomes unclickable, that's why I need to use reset filter function before adding another filter.

SAEb-ai replied:
Can you please share a video of it? From above it looks like it is an issue with the UI, right? Also, earlier why it won't used to be a problem with the protractor code.

ShivamJhaa replied:
No its not an issue with frontend. In webdriverio when we want to click on an element, webdriverio clicks at the center of that element. But in this case centre of the element to be clicked is covered with already assigned topic we cannot click and hence it throws error. In protractor click commands works different and it basically clicks on the particular portion of element which is clickable which is not the case with webdriverio.

https://user-images.githubusercontent.com/86483059/183242072-b861c565-daba-4c4f-9678-b9e1bb84374b.mp4

If we will look at the video, in suite file we have deleteTopicWithName at step 100(timestamp 0:36) where we first need to filter the skills on the basis of topic name but as we can see a topic name is already present in filter button we cant click there and it throws the error.

SAEb-ai replied:
As the keyword is already present, then why we need to filter it again?

ShivamJhaa replied:
The keyword present is not for the topic that we are going to delete in next step. In deleteTopicWithName the first step is to filter the topic with topic name provided. So we need to make sure there is no keyword already present in filter button.

SAEb-ai commented on core/tests/webdriverio_desktop/topicsAndSkillsDashboard.js:
```diff
@@ -126,8 +127,8 @@ describe('Topics and skills dashboard functionality', function() {
     let topicsCount = await topicsAndSkillsDashboardPage.getTopicsCount();
-    await topicsAndSkillsDashboardPage.filterTopicsByKeyword(TOPIC_ALPHA);
-    await topicsAndSkillsDashboardPage.filterTopicsByKeyword(TOPIC_BETA);
+    await topicsAndSkillsDashboardPage.filterTopicsByKeyword(
+      TOPIC_ALPHA + '\n' + ' ' + TOPIC_BETA);
```
Why this change?

ShivamJhaa replied:
As mentioned above, we can't click on the topic filter button if the filter is already applied, hence to use two filter keyword at a same time, I am passing both keywords in a single command.

SAEb-ai replied:
Ditto above.

ShivamJhaa replied:
Same as above, we need to pass two option for the filter button but as soon as we pass the first option the button becomes unclikable at the center, due to which I need to pass both the options in a single step only.

SAEb-ai replied:
What is this `' '` used for here?

ShivamJhaa replied:
Removed.

SAEb-ai commented on core/tests/webdriverio_utils/ClassroomPage.js:
```diff
@@ -0,0 +1,44 @@
+    var topicSummaryTiles = await topicSummaryTilesSelector();
+    if (count > 0) {
+      await waitFor.visibilityOf(
+        topicSummaryTiles[0], 'Topic summary tile is not visible');
```
Use a single selector for this.

ShivamJhaa replied:
Done.

SAEb-ai commented on core/tests/webdriverio_utils/LibraryPage.js:
```diff
@@ -101,31 +177,39 @@ var LibraryPage = function() {
 
+  this.getExplorationObjective = async function(name) {
+    var elems = await _getExplorationElements(name);
+    var value = await action.getText(
```
Can we not have a meaningful name for it other than value?

ShivamJhaa replied:
Done.

SAEb-ai replied:
Not Done

ShivamJhaa replied:
Done.

SAEb-ai commented on core/tests/webdriverio_utils/TopicsAndSkillsDashboardPage.js:
```diff
@@ -134,7 +135,7 @@ var TopicsAndSkillsDashboardPage = function() {
       'Merge skill button', mergeSkillsButton);
 
-    var skill = $(`.e2e-test-skills-list-item=${newSkillName}`);
+    var skill = $(`.e2e-test-skills-list-item*=${newSkillName}`);
```
Why was this change?

ShivamJhaa replied:
We do not pass the exact skill name that is present as text, instead we pass partial skill name text and `=` operators throws error if the string passed is not the exact string that is present, so I am using `*=` which is a partial string matching operator.

SAEb-ai commented on core/tests/webdriverio_utils/TopicsAndSkillsDashboardPage.js:
```diff
@@ -27,6 +27,7 @@ var TopicsAndSkillsDashboardPage = function() {
   var assignSkillToTopicButtonsSelector = function() {
     return $$('.e2e-test-assign-skill-to-topic-button');
   };
+  var assignedTopicNameInputClass = '.e2e-test-unassign-topic';
```
Why this was added?

ShivamJhaa replied:
It was not added by mistake when I migrated this dependency, it is same as we have in protractor. PTAL [here](https://github.com/oppia/oppia/blob/develop/core/tests/protractor_utils/TopicsAndSkillsDashboardPage.js#L109)

SAEb-ai replied:
If it was not added earlier, then why the tests were not failing earlier?

ShivamJhaa replied:
This selector is used in one of the test suite migrated in this PR, earlier it was not used any where that's why test were passing.

ShivamJhaa committed:
Added soln for element not defined flake

ShivamJhaa committed:
reverted few change

ShivamJhaa committed:
reverted few change

ShivamJhaa committed:
merge conflicts

ShivamJhaa committed:
few fixes

SAEb-ai commented on core/tests/webdriverio_utils/forms.js:
```diff
@@ -439,6 +439,8 @@ var MultiSelectEditor = function(elem) {
     },
     expectCurrentSelectionToBe: async function(expectedCurrentSelection) {
       // Open the dropdown menu.
+      var searchBarDropdownToggleElement = elem.$(
+        '.e2e-test-search-bar-dropdown-toggle');
```
Why was this added?

ShivamJhaa replied:
It was not added by mistake when I migrated this dependency, it is same as we have in protractor. PTAL [here](https://github.com/oppia/oppia/blob/develop/core/tests/protractor_utils/forms.js#L395)

SAEb-ai replied:
Ditto above.

ShivamJhaa replied:
Same as above.

SAEb-ai commented on core/tests/webdriverio_utils/workflow.js:
```diff
@@ -226,11 +226,11 @@ var _addExplorationRole = async function(roleName, username) {
 
 var addExplorationManager = async function(username) {
-  await _addExplorationRole('Manager', username);
+  await _addExplorationRole('Manager (can edit permissions)', username);
```
Why  (can edit permissions)? Ditto elsewhere?

ShivamJhaa replied:
As mentioned previously `selectByVisibleText` doesn't work if we do not pass the exact text that is present inside the `<option>` tag, so I am passing the exact string that we actually have in frontend.

SAEb-ai commented on core/tests/webdriverio_utils/workflow.js:
```diff
@@ -249,6 +249,18 @@ var addExplorationPlaytester = async function(username) {
 
+// Here, roleName is the server-side form of the name (e.g. 'owner').
+var _getExplorationRoles = async function(roleName) {
+  var ownerLocator = '.e2e-test-role-' + roleName + '-name';
```
Can you please explain this line because it seems to be a bit different from the protractor code?

ShivamJhaa replied:
As the page which is being tested here is still in Angular JS, protractor was using `by.repeater` but as there is no replacement for this in webdriverio, I just added a classname and used all selector for getting the text.

SAEb-ai replied:
Can we not have a more meaningful name here in terms of class?

ShivamJhaa replied:
Done.

SAEb-ai replied:
But still `explorationOwnerLocator` name doesn't seem to be related to the roles part.

ShivamJhaa replied:
Done.

SAEb-ai replied:
`explorationRoleNameLocator` would have been better

ShivamJhaa replied:
Done.

ShivamJhaa committed:
resolved comments

U8NWXD reviewed (approved):
Approving now that SAEb-ai has reviewed

ShivamJhaa committed:
Merge branch 'develop' of https://github.com/oppia/oppia into Milestone-1.7

ShivamJhaa committed:
resolved comments

ShivamJhaa committed:
Replace chromedriver service with selenium standalone service

SAEb-ai commented:
@ShivamJhaa Sorry for the late review. I have resolved some of the comments which I think have been resolved fully. PTAL. Thanks!!

ShivamJhaa committed:
Resolved comments

ShivamJhaa committed:
Fixed failing test

ShivamJhaa committed:
removed video and protractor upload workflow

ShivamJhaa commented:
I have added the solution for the [ECONRESET](https://github.com/oppia/oppia/runs/7369849467?check_suite_focus=true) flake in this PR. 
20 successful Cl run: https://github.com/oppia/oppia/runs/7729349300?check_suite_focus=true

This flake was coming due to the connection lost with the chromedriver backend server. I was not able to get to the root cause of this error but as I replaced chromedriver service with selenium standalone service to run the webdriverio tests, I was successfully able to resolve the flake. The changes related to the solution of this flake is [here](https://github.com/oppia/oppia/blob/f12d6e1001bced2294edffec71ee6c49d24662f2/core/tests/wdio.conf.js#L190).

SAEb-ai commented:
@ShivamJhaa Congrats on fixing the most occuring flake for milestone one. I just have two more comments left(last). I will approv it then.

ShivamJhaa committed:
Resolved comments

ShivamJhaa committed:
Merge branch 'develop' of https://github.com/oppia/oppia into Milestone-1.7

ShivamJhaa commented:
> @ShivamJhaa Congrats on fixing the most occuring flake for milestone one. I just have two more comments left(last). I will approv it then.

Thanks, I have addressed your comments. PTAL

SAEb-ai reviewed (approved):
Everything looks good to me. Except [this](https://github.com/oppia/oppia/runs/7750643809?check_suite_focus=true#step:13:347) flake. Do you know why this occured yesterday? Also, can this be because of selenium standalone?

ShivamJhaa commented:
> Everything looks good to me. Except [this](https://github.com/oppia/oppia/runs/7750643809?check_suite_focus=true#step:13:347) flake. Do you know why this occured yesterday? Also, can this be because of selenium standalone?

Extension suite is not yet migrated to webdriverio, so I don't think its because of the changes of this PR.
